pkg/expr/threshold.go

```diff
@@ -0,0 +1,134 @@
+package expr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/expr/mathexp"
+)
+
+type ThresholdCommand struct {
+	ReferenceVar  string
+	RefID         string
+	ThresholdFunc string
+	Conditions    []float64
+}
+
+const (
+	ThresholdIsAbove        = "gt"
+	ThresholdIsBelow        = "lt"
+	ThresholdIsWithinRange  = "within_range"
+	ThresholdIsOutsideRange = "outside_range"
+)
+
+func NewThresholdCommand(refID, referenceVar, thresholdFunc string, conditions []float64) (*ThresholdCommand, error) {
+	return &ThresholdCommand{
+		RefID:         refID,
+		ReferenceVar:  referenceVar,
+		ThresholdFunc: thresholdFunc,
+		Conditions:    conditions,
+	}, nil
+}
+
+type ThresholdConditionJSON struct {
+	Evaluator ConditionEvalJSON `json:"evaluator"`
+}
+
+type ConditionEvalJSON struct {
+	Params []float64 `json:"params"`
+	Type   string    `json:"type"` // e.g. "gt"
+}
+
+// UnmarshalResampleCommand creates a ResampleCMD from Grafana's frontend query.
+func UnmarshalThresholdCommand(rn *rawNode) (*ThresholdCommand, error) {
+	rawQuery := rn.Query
+
+	rawExpression, ok := rawQuery["expression"]
+	if !ok {
+		return nil, fmt.Errorf("no variable specified to reference for refId %v", rn.RefID)
+	}
+	referenceVar, ok := rawExpression.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected threshold variable to be a string, got %T for refId %v", rawExpression, rn.RefID)
+	}
+
+	jsonFromM, err := json.Marshal(rawQuery["conditions"])
+	if err != nil {
+		return nil, fmt.Errorf("failed to remarshal threshold expression body: %w", err)
+	}
+	var conditions []ThresholdConditionJSON
+	if err = json.Unmarshal(jsonFromM, &conditions); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal remarshaled threshold expression body: %w", err)
+	}
+
+	for _, condition := range conditions {
+		if !IsSupportedThresholdFunc(condition.Evaluator.Type) {
+			supportedThresholdFuncs := GetSupportedThresholdFuncs()
+			return nil, fmt.Errorf("expected threshold function to be one of %s, got %s", strings.Join(supportedThresholdFuncs, ", "), condition.Evaluator.Type)
+		}
+	}
+
+	// we only support one condition for now, we might want to turn this in to "OR" expressions later
+	if len(conditions) < 1 {
```

Comment:
If we just want to support one condition for now perhaps we should change this to `!=`?

```suggestion
	if len(conditions) != 1 {
```